backend/kesaseteli/applications/models.py

```diff
@@ -48,6 +50,69 @@ class Meta:
         ordering = ["name"]
 
 
+class YouthApplicationQuerySet(MatchesAnyOfMixin, models.QuerySet):
+    def _annotate_existence_duration(self):
+        """
+        Annotate queryset with existence_duration with value timezone.now() - created_at
+
+        :return: The source queryset with annotated existence_duration DurationField
+                 with value timezone.now() - created_at.
+        """
+        return self.annotate(
+            existence_duration=ExpressionWrapper(
+                timezone.now() - F("created_at"), output_field=DurationField()
+            )
+        )
+
+    def _active_q_filter(self) -> Q:
+        """
+        Return Q filter for active youth applications
+        """
+        return Q(receipt_confirmed_at__isnull=False)
+
+    def _unexpired_q_filter(self) -> Q:
+        """
+        Return Q filter for unexpired youth applications.
+
+        NOTE: Needs existence_duration field, so use with _annotate_existence_duration.
+        """
+        return Q(existence_duration__lt=YouthApplication.expiration_duration())
```

Comment:
Do you have any mechanism to auto clean expired YouthApplication? How many YouthApplication are we expected to have?
I think the way you query unexpired application is quite expensive and might cause performance issues when the number of applications grows up. Backend will always need to recalculate the existence duration of **every** Application in **every request** from the client. I believe an expired application most likely will never become unexpired again so it's not a good idea to try to re-calculate the existence duration of every application.
So, unless the system requires an application to be unavailable as soon as it expired, I'd make a separated BooleanField, to store the expired status of the application, then use a cronjob that runs every minute (or 5-15 minutes depends on the need) to query every unexpired application and update status if it expires

Reply:
No, there's no auto clean mechanism available at the moment.

Based on the previous years' data from a product owner I'd say the YouthApplication objects per year would probably realistically be in range 5k–10k. This is for one application per one person situation. As it is possible to submit multiple applications per person then this might make the resulting object count higher, e.g. double it to 20k.

It is true that the performance may become an issue with higher amount of YouthApplication objects as the `existence_duration` annotated by `YouthApplicationQuerySet._annotate_existence_duration` is not indexed and can't be indexed by a functional index because it would not be immutable.

We have the expiration time configurable by a shared setting between the backend and frontend to keep it dynamic. With this conversation it's evident this has a downside related to query performance. If we could and would hardcode the value then we could create a functional index on `YouthApplication.created_at + <hardcoded expiration time>` and then query against that using `timezone.now()` to filter out the expired YouthApplication objects. That query should hopefully be fast by using the functional index.

The marking with a cronjob sounds like a viable solution too.

I propose in the lean spirit to split the possible query performance issue fixing to another ticket and not include its implementation in this pull request. How would you feel about this?

Reply:
Totally agree, thanks for the clarification!

Reply:
You're welcome!